File: backend/src/rejection/rejection.service.ts
import { Injectable, Logger, NotFoundException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import {
  CandidateStatus,
  FeedbackAction,
  RejectionReason,
  Prisma,
} from '@prisma/client';
import { CreateRuleDto, RuleCondition, UpdateRuleDto } from './dto/feedback.dto';

interface CandidateSnapshot {
  totalScore: number | null;
  followers: number;
  publicRepos: number;
  totalCommits: number;
  company: string | null;
  location: string | null;
  primaryLanguage: string | null;
  sources: string[];
}

export interface RejectionStats {
  totalRejected: number;
  totalShortlisted: number;
  activeRules: number;
  reasonDistribution: { reason: RejectionReason; count: number; percentage: number }[];
  recentRejections: number;
}

@Injectable()
export class RejectionService {
  private readonly logger = new Logger(RejectionService.name);

  constructor(private prisma: PrismaService) {}

  async rejectCandidate(
    candidateId: string,
    reason: RejectionReason,
    notes?: string,
  ) {
    const candidate = await this.prisma.candidate.findUnique({
      where: { id: candidateId },
      include: {
        sources: true,
        repositories: { take: 1, orderBy: { starCount: 'desc' } },
      },
    });

    if (!candidate) {
      throw new NotFoundException(`Candidate ${candidateId} not found`);
    }

    const snapshot: CandidateSnapshot = {
      totalScore: candidate.totalScore,
      followers: candidate.followers,
      publicRepos: candidate.publicRepos,
      totalCommits: candidate.totalCommits,
      company: candidate.company,
      location: candidate.location,
      primaryLanguage: candidate.repositories[0]?.language || null,
      sources: candidate.sources.map((s) => s.sourceName),
    };

    await this.prisma.$transaction([
      this.prisma.candidate.update({
        where: { id: candidateId },
        data: { status: CandidateStatus.REJECTED },
      }),
      this.prisma.candidateFeedback.create({
        data: {
          candidateId,
          action: FeedbackAction.REJECT,
          reason,
          notes,
          snapshot: snapshot as unknown as Prisma.InputJsonValue,
        },
      }),
    ]);

    this.logger.log(`Candidate ${candidateId} rejected with reason: ${reason}`);

    return { success: true, status: CandidateStatus.REJECTED };
  }

  async shortlistCandidate(candidateId: string) {
    const candidate = await this.prisma.candidate.findUnique({
      where: { id: candidateId },
    });

    if (!candidate) {
      throw new NotFoundException(`Candidate ${candidateId} not found`);
    }

    await this.prisma.$transaction([
      this.prisma.candidate.update({
        where: { id: candidateId },
        data: { status: CandidateStatus.SHORTLISTED },
      }),
      this.prisma.candidateFeedback.create({
        data: {
          candidateId,
          action: FeedbackAction.SHORTLIST,
        },
      }),
    ]);

    this.logger.log(`Candidate ${candidateId} shortlisted`);

    return { success: true, status: CandidateStatus.SHORTLISTED };
  }

  async undoFeedback(candidateId: string) {
    const candidate = await this.prisma.candidate.findUnique({
      where: { id: candidateId },
    });

    if (!candidate) {
      throw new NotFoundException(`Candidate ${candidateId} not found`);
    }

    await this.prisma.$transaction([
      this.prisma.candidate.update({
        where: { id: candidateId },
        data: { status: CandidateStatus.ACTIVE },
      }),
      this.prisma.candidateFeedback.create({
        data: {
          candidateId,
          action: FeedbackAction.UNDO,
        },
      }),
    ]);

    this.logger.log(`Candidate ${candidateId} feedback undone`);

    return { success: true, status: CandidateStatus.ACTIVE };
  }

  async getStats(): Promise<RejectionStats> {
    const [
      totalRejected,
      totalShortlisted,
      activeRules,
      reasonCounts,
      recentRejections,
    ] = await Promise.all([
      this.prisma.candidate.count({
        where: { status: CandidateStatus.REJECTED },
      }),
      this.prisma.candidate.count({
        where: { status: CandidateStatus.SHORTLISTED },
      }),
      this.prisma.rejectionRule.count({
        where: { enabled: true },
      }),
      this.prisma.candidateFeedback.groupBy({
        by: ['reason'],
        where: {
          action: FeedbackAction.REJECT,
          reason: { not: null },
        },
        _count: { reason: true },
      }),
      this.prisma.candidateFeedback.count({
        where: {
          action: FeedbackAction.REJECT,
          createdAt: { gte: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000) },
        },
      }),
    ]);

    const totalReasons = reasonCounts.reduce(
      (sum, r) => sum + r._count.reason,
      0,
    );

    const reasonDistribution = reasonCounts
      .filter((r) => r.reason !== null)
      .map((r) => ({
        reason: r.reason as RejectionReason,
        count: r._count.reason,
        percentage:
          totalReasons > 0
            ? Math.round((r._count.reason / totalReasons) * 100)
            : 0,
      }))
      .sort((a, b) => b.count - a.count);

    return {
      totalRejected,
      totalShortlisted,
      activeRules,
      reasonDistribution,
      recentRejections,
    };
  }

  async getRules() {
    return this.prisma.rejectionRule.findMany({
      orderBy: [{ enabled: 'desc' }, { confidence: 'desc' }],
    });
  }

  async createRule(dto: CreateRuleDto) {
    return this.prisma.rejectionRule.create({
      data: {
        name: dto.name,
        description: dto.description,
        conditions: dto.conditions as unknown as Prisma.InputJsonValue,
        autoGenerated: false,
      },
    });
  }

  async updateRule(id: string, dto: UpdateRuleDto) {
    const rule = await this.prisma.rejectionRule.findUnique({
      where: { id },
    });

    if (!rule) {
      throw new NotFoundException(`Rule ${id} not found`);
    }

    return this.prisma.rejectionRule.update({
      where: { id },
      data: {
        ...(dto.name && { name: dto.name }),
        ...(dto.description !== undefined && { description: dto.description }),
        ...(dto.conditions && {
          conditions: dto.conditions as unknown as Prisma.InputJsonValue,
        }),
        ...(dto.enabled !== undefined && { enabled: dto.enabled }),
      },
    });
  }

  async deleteRule(id: string) {
    const rule = await this.prisma.rejectionRule.findUnique({
      where: { id },
    });

    if (!rule) {
      throw new NotFoundException(`Rule ${id} not found`);
    }

    await this.prisma.rejectionRule.delete({ where: { id } });

    return { success: true };
  }

  async checkAutoExclude(candidateId: string): Promise<{
    shouldExclude: boolean;
    matchedRules: string[];
  }> {
    const candidate = await this.prisma.candidate.findUnique({
      where: { id: candidateId },
      include: {
        sources: true,
        repositories: { take: 1, orderBy: { starCount: 'desc' } },
      },
    });

    if (!candidate) {
      return { shouldExclude: false, matchedRules: [] };
    }

    const enabledRules = await this.prisma.rejectionRule.findMany({
      where: { enabled: true },
    });

    const matchedRules: string[] = [];

    for (const rule of enabledRules) {
      const conditions = rule.conditions as unknown as RuleCondition[];
      const matches = this.evaluateConditions(candidate, conditions);

      if (matches) {
        matchedRules.push(rule.name);
      }
    }

    return {
      shouldExclude: matchedRules.length > 0,
      matchedRules,
    };
  }

  private evaluateConditions(
    candidate: {
      totalScore: number | null;
      followers: number;
      publicRepos: number;
      totalCommits: number;
      company: string | null;
      location: string | null;
      sources: { sourceName: string }[];
      repositories: { language: string | null }[];
    },
    conditions: RuleCondition[],
  ): boolean {
    for (const condition of conditions) {
      const value = this.getCandidateFieldValue(candidate, condition.field);
      const matches = this.evaluateCondition(value, condition);

      if (!matches) {
        return false;
      }
    }

    return conditions.length > 0;
  }

  private getCandidateFieldValue(
    candidate: {
      totalScore: number | null;
      followers: number;
      publicRepos: number;
      totalCommits: number;
      company: string | null;
      location: string | null;
      sources: { sourceName: string }[];
      repositories: { language: string | null }[];
    },
    field: string,
  ): unknown {
    switch (field) {
      case 'totalScore':
        return candidate.totalScore;
      case 'followers':
        return candidate.followers;
      case 'publicRepos':
        return candidate.publicRepos;
      case 'totalCommits':
        return candidate.totalCommits;
      case 'company':
        return candidate.company;
      case 'location':
        return candidate.location;
      case 'primaryLanguage':
        return candidate.repositories[0]?.language;
      case 'sources':
        return candidate.sources.map((s) => s.sourceName);
      default:
        return undefined;
    }
  }

  private evaluateCondition(value: unknown, condition: RuleCondition): boolean {
    const { operator, value: conditionValue } = condition;

    if (value === null || value === undefined) {
      return operator === '=' && conditionValue === null;
    }

    switch (operator) {
      case '<':
        return (value as number) < (conditionValue as number);
      case '>':
        return (value as number) > (conditionValue as number);
      case '<=':
        return (value as number) <= (conditionValue as number);
      case '>=':
        return (value as number) >= (conditionValue as number);
      case '=':
        return value === conditionValue;
      case '!=':
        return value !== conditionValue;
      case 'in':
        return (conditionValue as unknown[]).includes(value);
      case 'notIn':
        return !(conditionValue as unknown[]).includes(value);
      case 'contains':
        if (Array.isArray(value)) {
          return value.includes(conditionValue);
        }
        return String(value).includes(String(conditionValue));
      default:
        return false;
    }
  }
}

File: backend-java/src/main/java/com/peopleminer/rejection/RejectionService.java
package com.peopleminer.rejection;

import com.peopleminer.domain.entity.Candidate;
import com.peopleminer.domain.entity.CandidateFeedback;
import com.peopleminer.domain.entity.RejectionRule;
import com.peopleminer.domain.enums.CandidateStatus;
import com.peopleminer.domain.enums.FeedbackAction;
import com.peopleminer.domain.enums.RejectionReason;
import com.peopleminer.domain.repository.CandidateFeedbackRepository;
import com.peopleminer.domain.repository.CandidateRepository;
import com.peopleminer.domain.repository.RejectionRuleRepository;
import com.peopleminer.rejection.dto.CreateRuleRequest;
import com.peopleminer.rejection.dto.UpdateRuleRequest;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class RejectionService {

    private final CandidateRepository candidateRepository;
    private final CandidateFeedbackRepository feedbackRepository;
    private final RejectionRuleRepository ruleRepository;

    @Transactional
    public Map<String, Object> rejectCandidate(String candidateId, RejectionReason reason, String notes) {
        Candidate candidate = candidateRepository.findByIdWithRelations(candidateId)
            .orElseThrow(() -> new EntityNotFoundException("Candidate " + candidateId + " not found"));

        Map<String, Object> snapshot = new HashMap<>();
        snapshot.put("totalScore", candidate.getTotalScore());
        snapshot.put("followers", candidate.getFollowers());
        snapshot.put("publicRepos", candidate.getPublicRepos());
        snapshot.put("totalCommits", candidate.getTotalCommits());
        snapshot.put("company", candidate.getCompany());
        snapshot.put("location", candidate.getLocation());
        snapshot.put("primaryLanguage", !candidate.getRepositories().isEmpty()
            ? candidate.getRepositories().get(0).getLanguage() : null);
        snapshot.put("sources", candidate.getSources().stream()
            .map(s -> s.getSourceName())
            .toList());

        candidate.setStatus(CandidateStatus.REJECTED);
        candidateRepository.save(candidate);

        CandidateFeedback feedback = CandidateFeedback.builder()
            .action(FeedbackAction.REJECT)
            .reason(reason)
            .notes(notes)
            .snapshot(snapshot)
            .build();
        candidate.addFeedback(feedback);
        feedbackRepository.save(feedback);

        log.info("Candidate {} rejected with reason: {}", candidateId, reason);

        return Map.of(
            "success", true,
            "status", CandidateStatus.REJECTED.name()
        );
    }

    @Transactional
    public Map<String, Object> shortlistCandidate(String candidateId) {
        Candidate candidate = candidateRepository.findById(candidateId)
            .orElseThrow(() -> new EntityNotFoundException("Candidate " + candidateId + " not found"));

        candidate.setStatus(CandidateStatus.SHORTLISTED);
        candidateRepository.save(candidate);

        CandidateFeedback feedback = CandidateFeedback.builder()
            .action(FeedbackAction.SHORTLIST)
            .build();
        candidate.addFeedback(feedback);
        feedbackRepository.save(feedback);

        log.info("Candidate {} shortlisted", candidateId);

        return Map.of(
            "success", true,
            "status", CandidateStatus.SHORTLISTED.name()
        );
    }

    @Transactional
    public Map<String, Object> undoFeedback(String candidateId) {
        Candidate candidate = candidateRepository.findById(candidateId)
            .orElseThrow(() -> new EntityNotFoundException("Candidate " + candidateId + " not found"));

        candidate.setStatus(CandidateStatus.ACTIVE);
        candidateRepository.save(candidate);

        CandidateFeedback feedback = CandidateFeedback.builder()
            .action(FeedbackAction.UNDO)
            .build();
        candidate.addFeedback(feedback);
        feedbackRepository.save(feedback);

        log.info("Candidate {} feedback undone", candidateId);

        return Map.of(
            "success", true,
            "status", CandidateStatus.ACTIVE.name()
        );
    }

    @Transactional(readOnly = true)
    public RejectionStats getStats() {
        long totalRejected = candidateRepository.countByStatus(CandidateStatus.REJECTED);
        long totalShortlisted = candidateRepository.countByStatus(CandidateStatus.SHORTLISTED);
        long activeRules = ruleRepository.countByEnabledTrue();

        List<Object[]> reasonCounts = feedbackRepository.countByReason();
        long totalReasons = reasonCounts.stream()
            .mapToLong(row -> (Long) row[1])
            .sum();

        List<RejectionStats.ReasonCount> reasonDistribution = reasonCounts.stream()
            .map(row -> {
                RejectionReason reason = (RejectionReason) row[0];
                long count = (Long) row[1];
                int percentage = totalReasons > 0 ? (int) Math.round(count * 100.0 / totalReasons) : 0;
                return RejectionStats.ReasonCount.builder()
                    .reason(reason)
                    .count(count)
                    .percentage(percentage)
                    .build();
            })
            .sorted(Comparator.comparingLong(RejectionStats.ReasonCount::getCount).reversed())
            .collect(Collectors.toList());

        LocalDateTime weekAgo = LocalDateTime.now().minusDays(7);
        long recentRejections = feedbackRepository.countByActionAndCreatedAtGreaterThanEqual(
            FeedbackAction.REJECT, weekAgo);

        return RejectionStats.builder()
            .totalRejected(totalRejected)
            .totalShortlisted(totalShortlisted)
            .activeRules(activeRules)
            .reasonDistribution(reasonDistribution)
            .recentRejections(recentRejections)
            .build();
    }

    @Transactional(readOnly = true)
    public List<RejectionRule> getRules() {
        return ruleRepository.findAllByOrderByEnabledDescConfidenceDesc();
    }

    @Transactional
    public RejectionRule createRule(CreateRuleRequest request) {
        RejectionRule rule = RejectionRule.builder()
            .name(request.getName())
            .description(request.getDescription())
            .conditions(request.getConditions())
            .autoGenerated(false)
            .build();
        return ruleRepository.save(rule);
    }

    @Transactional
    public RejectionRule updateRule(String id, UpdateRuleRequest request) {
        RejectionRule rule = ruleRepository.findById(id)
            .orElseThrow(() -> new EntityNotFoundException("Rule " + id + " not found"));

        if (request.getName() != null) {
            rule.setName(request.getName());
        }
        if (request.getDescription() != null) {
            rule.setDescription(request.getDescription());
        }
        if (request.getConditions() != null) {
            rule.setConditions(request.getConditions());
        }
        if (request.getEnabled() != null) {
            rule.setEnabled(request.getEnabled());
        }

        return ruleRepository.save(rule);
    }

    @Transactional
    public Map<String, Object> deleteRule(String id) {
        RejectionRule rule = ruleRepository.findById(id)
            .orElseThrow(() -> new EntityNotFoundException("Rule " + id + " not found"));

        ruleRepository.delete(rule);
        return Map.of("success", true);
    }

    @Transactional(readOnly = true)
    public Map<String, Object> checkAutoExclude(String candidateId) {
        Candidate candidate = candidateRepository.findByIdWithRelations(candidateId)
            .orElse(null);

        if (candidate == null) {
            return Map.of("shouldExclude", false, "matchedRules", List.of());
        }

        List<RejectionRule> enabledRules = ruleRepository.findByEnabledTrueOrderByConfidenceDesc();
        List<String> matchedRules = new ArrayList<>();

        for (RejectionRule rule : enabledRules) {
            if (evaluateConditions(candidate, rule.getConditions())) {
                matchedRules.add(rule.getName());
            }
        }

        return Map.of(
            "shouldExclude", !matchedRules.isEmpty(),
            "matchedRules", matchedRules
        );
    }

    private boolean evaluateConditions(Candidate candidate, List<RejectionRule.RuleCondition> conditions) {
        if (conditions == null || conditions.isEmpty()) {
            return false;
        }

        for (RejectionRule.RuleCondition condition : conditions) {
            Object value = getCandidateFieldValue(candidate, condition.getField());
            if (!evaluateCondition(value, condition)) {
                return false;
            }
        }

        return true;
    }

    private Object getCandidateFieldValue(Candidate candidate, String field) {
        return switch (field) {
            case "totalScore" -> candidate.getTotalScore();
            case "followers" -> candidate.getFollowers();
            case "publicRepos" -> candidate.getPublicRepos();
            case "totalCommits" -> candidate.getTotalCommits();
            case "company" -> candidate.getCompany();
            case "location" -> candidate.getLocation();
            case "primaryLanguage" -> candidate.getRepositories().isEmpty() ? null
                : candidate.getRepositories().get(0).getLanguage();
            case "sources" -> candidate.getSources().stream()
                .map(s -> s.getSourceName())
                .toList();
            default -> null;
        };
    }

    @SuppressWarnings("unchecked")
    private boolean evaluateCondition(Object value, RejectionRule.RuleCondition condition) {
        String operator = condition.getOperator();
        Object conditionValue = condition.getValue();

        if (value == null) {
            return "=".equals(operator) && conditionValue == null;
        }

        return switch (operator) {
            case "<" -> compareNumbers(value, conditionValue) < 0;
            case ">" -> compareNumbers(value, conditionValue) > 0;
            case "<=" -> compareNumbers(value, conditionValue) <= 0;
            case ">=" -> compareNumbers(value, conditionValue) >= 0;
            case "=" -> Objects.equals(value, conditionValue);
            case "!=" -> !Objects.equals(value, conditionValue);
            case "in" -> conditionValue instanceof List && ((List<?>) conditionValue).contains(value);
            case "notIn" -> !(conditionValue instanceof List && ((List<?>) conditionValue).contains(value));
            case "contains" -> {
                if (value instanceof List) {
                    yield ((List<?>) value).contains(conditionValue);
                }
                yield String.valueOf(value).contains(String.valueOf(conditionValue));
            }
            default -> false;
        };
    }

    private int compareNumbers(Object a, Object b) {
        double numA = a instanceof Number ? ((Number) a).doubleValue() : 0;
        double numB = b instanceof Number ? ((Number) b).doubleValue() : 0;
        return Double.compare(numA, numB);
    }
}
